lib/charms/alertmanager_k8s/v0/consumer.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+""" # AlertmanagerConsumer library
+
+This library is design to be used by a charm consuming the alertmanager-k8s relation.
+"""
+
+
+import ops
+from ops.framework import StoredState, EventBase, EventSource
+from ops.relation import ConsumerBase
+
+from typing import List
+import logging
+
+LIBID = "abcdef1234"  # Unique ID that refers to the library forever
+LIBAPI = 0    # Must match the major version in the import path.
+LIBPATCH = 1  # The current patch version. Must be updated when changing.
+
+logger = logging.getLogger(__name__)
+
+
+class AlertmanagerConsumer(ConsumerBase):
+    """A "consumer" handler to be used by charms that relate to Alertmanager.
+    This consumer auto-registers relation events on behalf of the user and communicates information
+    directly via `_stored` TODO: have a documented contract and act on it in the "available" hook.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            relation_name (str): from consumer's metadata.yaml
+            consumes (dict): provider specifications
+            multi (bool): multiple relations flag
+
+    Attributes:
+            charm (CharmBase): consumer charm
+    """
+    _stored: StoredState
+
+    def __init__(self, charm: ops.charm.CharmBase, relation_name, consumes, multi=False):
+        super().__init__(charm, relation_name, consumes, multi)
+        self.charm = charm
+        self._consumer_relation_name = relation_name  # from consumer's metadata.yaml
+        # self._provider_relation_name = "alerting"  # from alertmanager's metadata.yaml
+
+        self.framework.observe(self.charm.on[self._consumer_relation_name].relation_changed,
+                               self._on_relation_changed)
+        self.framework.observe(self.charm.on[self._consumer_relation_name].relation_departed,
+                               self._on_relation_departed)
+        self.framework.observe(self.charm.on[self._consumer_relation_name].relation_broken,
+                               self._on_relation_broken)
+
+        self._stored.set_default(alertmanagers={})
+
+    def _on_relation_changed(self, event: ops.charm.RelationChangedEvent):
+        """This hook stores locally the address of the newly-joined alertmanager.
+        This is needed for consumers such as prometheus, which should be aware of all alertmanager
+        instances.
+        """
+        if event.unit:  # event.unit may be `None` in the case of app data change
+            # Save locally the public IP address of the alertmanager unit
+            if address := event.relation.data[event.unit].get("public_address"):
+                # TODO consider storing in unit data instead of StoredState
+                self._stored.alertmanagers[event.unit.name] = address
+
+                # inform consumer about the change
+                self.on.available.emit()
```

Comment:
Doesn't this mean there will be two `available` events emitted for each joined unit (since the ConsumerBase also emits one) ? Is that the intention ?  Instead would it be possible to just ensure that `get_cluster_info()` provides the required list of addresses to Prometheus when `Prometheus.AlertmanagerConsumer` raises `available` ?

Reply:
Of course Prometheus will have to "pull" from consumer lib using `get_cluster_info()`. I mean `ConsumerBase` emits `ProviderAvailable` if the provider and consumer are compatible. How does this check for compatibility factor into the workflow being proposed in this PR. Can it happen that `ConsumerBase` emits `ProviderInvalid` but `AlertmanagerConsumer` emits `ProviderAvailable` in response to the same relation changed event ?

Reply:
That is a very good point.
That seems to suggest we should consider a custom event after all?

Reply:
Custom event would certainly be one way. I am always looking for the "simplest" way. I have not dug deep into this as I trust you are in a better position to do that and trust your judgement. But if you would like to keep things simple kindly do consider if it is possible that the `AlertmanagerConsumer` charm library exposes just one function `get_cluster_info()` and nothing else (no event handler). Since Prometheus just needs the list of alertmangers any time one of the `ConsumerBase` events fire.

Reply:
I tested without manually emitting, and occasionally there are misses. I'll keep it in for now until the `ready` flag on scale-up/down is resolved.

Reply:
I do not see how `ready` flag will solve the issue being raised here.

Reply:
Resolving conversation as a custom event is now used.